gc-worker/server/server.go

```diff
@@ -0,0 +1,299 @@
+	etcdClient *clientv3.Client
+	pdClient   pd.Client
+}
+
+// CreateServer creates the UNINITIALIZED pd server with given configuration.
+func CreateServer(ctx context.Context, cfg *Config) (*Server, error) {
+	log.Info("GCWorker Config", zap.Reflect("config", cfg))
+
+	s := &Server{
+		cfg:            cfg,
+		ctx:            ctx,
+		startTimestamp: time.Now().Unix(),
+	}
+
+	err := s.createPdClient(ctx)
+	if err != nil {
+		log.Error("create pd client fail", zap.Error(err))
+		return nil, err
+	}
+
+	err = s.createEtcdClient()
+	if err != nil {
+		log.Error("create pd client fail", zap.Error(err))
+		return nil, err
+	}
+	return s, nil
+}
+
+func (s *Server) GetServerName() string {
+	return s.cfg.Name
+}
+
+func (s *Server) createPdClient(ctx context.Context) error {
+	addrs := strings.Split(s.cfg.PdAddrs, ",")
+	securityOption := pd.SecurityOption{}
+	maxCallMsgSize := []grpc.DialOption{
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)),
+		grpc.WithDefaultCallOptions(grpc.MaxCallSendMsgSize(maxMsgSize)),
+	}
+	pdClient, err := pd.NewClientWithContext(
+		ctx, addrs, securityOption,
+		pd.WithGRPCDialOptions(maxCallMsgSize...),
+		pd.WithCustomTimeoutOption(10*time.Second),
+		pd.WithMaxErrorRetry(3),
+	)
+	if err != nil {
+		log.Error("fail to create pd client", zap.Error(err))
+		return errors.Trace(err)
+	}
+	s.pdClient = pdClient
+	return nil
+}
+
+func (s *Server) createEtcdClient() error {
+	endpoints := strings.Split(s.cfg.EtcdEndpoint, ",")
+	log.Info("create etcd v3 client", zap.Strings("endpoints", endpoints), zap.Reflect("cert", s.cfg.Security))
+
+	lgc := zap.NewProductionConfig()
+	lgc.Encoding = log.ZapEncodingName
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: etcdTimeout,
+		TLS:         nil,
+		LogConfig:   &lgc,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	s.etcdClient = client
+	return nil
+}
+
+// Close closes the server.
+func (s *Server) Close() {
+	if s.etcdClient != nil {
+		if err := s.etcdClient.Close(); err != nil {
+			log.Error("close etcd client meet error", zap.Error(err))
+		}
+	}
+	if s.pdClient != nil {
+		s.pdClient.Close()
+	}
+
+	if !atomic.CompareAndSwapInt64(&s.isServing, 1, 0) {
+		// server is already closed
+		return
+	}
+
+	log.Info("closing server")
+	s.stopServerLoop()
+
+	log.Info("closed server")
+}
+
+// checks whether server is closed or not.
+func (s *Server) IsServing() bool {
+	return atomic.LoadInt64(&s.isServing) != 0
+}
+
+func (s *Server) IsLead() bool {
+	return atomic.LoadInt64(&s.isLead) != 0
+}
+
+func (s *Server) StartServer() {
+	atomic.StoreInt64(&s.isServing, 1)
+	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(s.ctx)
+	s.serverLoopWg.Add(2)
+	go s.startEtcdLoop()
+	go s.startUpdateGCSafePointLoop()
+}
+
+func (s *Server) stopServerLoop() {
+	s.serverLoopCancel()
+	s.serverLoopWg.Wait()
+}
+
+// ComposeTS creates a ts from physical and logical parts.
+func ComposeTS(physical, logical int64) uint64 {
```

Comment:
Suggest move these timestamp functions to a new "util.go" file.
Please also add unit tests for them.

Reply:
I copy these function from client-go reps. And i found PD tsoutils package has the same functions. So i changed to depend on PD tsoutil.